File: kakao.app/js/ked/edit/tool/pair.js
var _k_ = {empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, last: function (o) {return o != null ? o.length ? o[o.length-1] : undefined : o}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }}

var pair

import kxk from "../../../kxk.js"
let kstr = kxk.kstr
let kutil = kxk.kutil
let kseg = kxk.kseg

import pepe from "../../../kxk/pepe.js"

import belt from "./belt.js"


pair = (function ()
{
    function pair ()
    {}

    pair["isUnbalancedPosition"] = function (lines, pos, char)
    {
        var p, revs, start, _31_28_

        revs = {']':'[','}':'{',')':'(','"':'"',"'":"'"}
        p = pepe(kseg.str(lines[pos[1]]))
        start = (p.unbalanced != null ? p.unbalanced.slice(1).map(function (s)
        {
            return s.start
        }) : undefined)
        return !_k_.empty(start) && _k_.in(revs[char],start)
    }

    pair["isRangeInString"] = function (lines, rng)
    {
        var _40_76_

        return (this.rangeOfStringSurroundingRange(lines,rng) != null)
    }

    pair["rangeOfStringSurroundingRange"] = function (lines, rng)
    {
        var ir

        if (ir = this.rangeOfInnerStringSurroundingRange(lines,rng))
        {
            return this.rangeGrownBy(ir,1)
        }
    }

    pair["rangeOfInnerStringSurroundingRange"] = function (lines, rng)
    {
        var r, rgs

        if (this.isInvalidLineIndex(lines,rng[1]))
        {
            return
        }
        rgs = this.rangesOfStringsInText(lines[rng[1]],rng[1])
        rgs = this.rangesShrunkenBy(rgs,1)
        var list = _k_.list(rgs)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            r = list[_a_]
            if (this.rangeContainsRange(r,rng))
            {
                return r
            }
        }
    }

    pair["rangesOfStringsInText"] = function (text, li = 0)
    {
        var c, cc, i, rngs, ss

        rngs = []
        ss = -1
        cc = null
        for (var _a_ = i = 0, _b_ = text.length; (_a_ <= _b_ ? i < text.length : i > text.length); (_a_ <= _b_ ? ++i : --i))
        {
            c = text[i]
            if (!cc && _k_.in(c,"'\""))
            {
                cc = c
                ss = i
            }
            else if (c === cc)
            {
                if ((text[i - 1] !== '\\') || (i > 2 && text[i - 2] === '\\'))
                {
                    rngs.push([ss,li,i + 1,li])
                    cc = null
                    ss = -1
                }
            }
        }
        return rngs
    }

    pair["positionsAndRangesOutsideStrings"] = function (lines, rngs, posl)
    {
        var found, pos, rng

        found = []
        var list = _k_.list(rngs)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            rng = list[_a_]
            if (!belt.isRangeInString(lines,rng))
            {
                found.push(rng)
            }
        }
        var list1 = _k_.list(posl)
        for (var _b_ = 0; _b_ < list1.length; _b_++)
        {
            pos = list1[_b_]
            if (!belt.isRangeInString(lines,belt.rangeForPos(pos)))
            {
                found.push(pos)
            }
        }
        return found
    }

    pair["rangesOfPairsSurroundingPositions"] = function (lines, pairs, posl)
    {
        var pair, pos, rngs

        rngs = []
        var list = _k_.list(posl)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            pos = list[_a_]
            var list1 = _k_.list(pairs)
            for (var _b_ = 0; _b_ < list1.length; _b_++)
            {
                pair = list1[_b_]
                if (this.chunkBeforePos(lines,pos).endsWith(pair[0]) && this.chunkAfterPos(lines,pos).startsWith(pair[1]))
                {
                    rngs.push([pos[0] - pair[0].length,pos[1],pos[0] + pair[1].length,pos[1]])
                }
            }
        }
        return rngs
    }

    pair["stringDelimiterSpansForPositions"] = function (lines, posl)
    {
        var pos, spans, srng

        spans = []
        var list = _k_.list(posl)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            pos = list[_a_]
            if (srng = this.rangeOfStringSurroundingRange(lines,[pos[0],pos[1],pos[0],pos[1]]))
            {
                spans.push([srng[0],srng[1],srng[0] + 1])
                spans.push([srng[2] - 1,srng[3],srng[2]])
            }
            else if (_k_.in(lines[pos[1]][pos[0]],"\"'"))
            {
                if (srng = this.rangeOfStringSurroundingRange(lines,[pos[0] + 1,pos[1],pos[0] + 1,pos[1]]))
                {
                    spans.push([srng[0],srng[1],srng[0] + 1])
                    spans.push([srng[2] - 1,srng[3],srng[2]])
                }
            }
        }
        return spans
    }

    pair["openCloseSpansForPositions"] = function (lines, posl)
    {
        var pos, spans, sps

        spans = []
        var list = _k_.list(posl)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            pos = list[_a_]
            if (sps = this.openCloseSpansForPosition(lines,pos))
            {
                spans = spans.concat(sps)
            }
        }
        return spans
    }

    pair["openCloseSpansForPosition"] = function (lines, pos)
    {
        var ap, bp, clos, closeEncounters, cnt, firstClose, lastOpen, maxLookups, next, open, openEncounters, opns, prev, revs, stack

        open = {'[':']','{':'}','(':')'}
        opns = '[{('
        clos = ']})'
        revs = {']':'[','}':'{',')':'('}
        maxLookups = 1000
        bp = [pos[0],pos[1]]
        if (!(_k_.in(lines[bp[1]][bp[0]],opns)))
        {
            closeEncounters = ''
            openEncounters = ''
            stack = []
            cnt = 0
            while (true)
            {
                bp[0] -= 1
                if (bp[0] >= 0)
                {
                    prev = lines[bp[1]][bp[0]]
                    if (_k_.in(prev,opns))
                    {
                        if (stack.length)
                        {
                            if (open[prev] === _k_.last(stack))
                            {
                                openEncounters += prev
                                stack.pop()
                                continue
                            }
                            else
                            {
                                return
                            }
                        }
                        lastOpen = prev
                        break
                    }
                    else if (_k_.in(prev,clos))
                    {
                        stack.push(prev)
                        closeEncounters += prev
                    }
                }
                else
                {
                    bp[1] -= 1
                    if (bp[1] < 0)
                    {
                        break
                    }
                    bp[0] = lines[bp[1]].length
                }
                if ((lastOpen != null))
                {
                    break
                }
                if (bp[1] < 0)
                {
                    break
                }
                if (cnt++ > maxLookups)
                {
                    break
                }
            }
        }
        else
        {
            lastOpen = lines[bp[1]][bp[0]]
        }
        stack = []
        ap = [_k_.max(bp[0] + 1,pos[0]),pos[1]]
        cnt = 0
        while (ap[1] < lines.length)
        {
            next = lines[ap[1]][ap[0]]
            if (_k_.in(next,clos))
            {
                if (stack.length)
                {
                    if (open[_k_.last(stack)] === next)
                    {
                        stack.pop()
                    }
                    else
                    {
                        return
                    }
                }
                else
                {
                    firstClose = next
                    break
                }
            }
            else if (_k_.in(next,opns))
            {
                stack.push(next)
            }
            ap[0] += 1
            if (ap[0] >= lines[ap[1]].length)
            {
                ap[0] = 0
                ap[1] += 1
            }
            if (cnt++ > maxLookups)
            {
                break
            }
        }
        if (!(lastOpen != null) || !(firstClose != null))
        {
            if (_k_.in(lines[pos[1]][pos[0] - 1],clos) && _k_.in(revs[lines[pos[1]][pos[0] - 1]],openEncounters))
            {
                return this.openCloseSpansForPosition(lines,[pos[0] - 1,pos[1]])
            }
            return
        }
        if (open[lastOpen] === firstClose)
        {
            return [[bp[0],bp[1],bp[0] + 1],[ap[0],ap[1],ap[0] + 1]]
        }
    }

    return pair
})()

export default pair;

File: kakao.app/love/lua/edit/tool/pair.lua
--[[
000000000  00000000  000   000  000000000
   000     000        000 000      000   
   000     0000000     00000       000   
   000     000        000 000      000   
   000     00000000  000   000     000   
--]]


local pair = class("pair")
    

-- ███   ███  ███   ███  ███████     ███████   ███       ███████   ███   ███   ███████  ████████  ███████  
-- ███   ███  ████  ███  ███   ███  ███   ███  ███      ███   ███  ████  ███  ███       ███       ███   ███
-- ███   ███  ███ █ ███  ███████    █████████  ███      █████████  ███ █ ███  ███       ███████   ███   ███
-- ███   ███  ███  ████  ███   ███  ███   ███  ███      ███   ███  ███  ████  ███       ███       ███   ███
--  ███████   ███   ███  ███████    ███   ███  ███████  ███   ███  ███   ███   ███████  ████████  ███████  


function pair.static.isUnbalancedPosition(lines, pos, char) 
        -- revs = {
        --     ']': '[' 
        --     '}': '{' 
        --     ')': '('
        --     '"': '"'
        --     "'": "'"}
        --        
        -- p = pepe kseg.str(lines[pos[2]])
        -- start = p.unbalanced∙slice(2)∙map((s) -> s.start)
        -- valid start and start∙has(revs[char])
        return false
    end

--  0000000  000000000  00000000   000  000   000   0000000   
-- 000          000     000   000  000  0000  000  000        
-- 0000000      000     0000000    000  000 0 000  000  0000  
--      000     000     000   000  000  000  0000  000   000  
-- 0000000      000     000   000  000  000   000   0000000   


function pair.static.isRangeInString(lines, rng) 
    return (pair.rangeOfStringSurroundingRange(lines, rng) ~= nil)
    end


function pair.static.rangeOfStringSurroundingRange(lines, rng) 
        local ir = belt.rangeOfInnerStringSurroundingRange(lines, rng)
        if ir then 
            return belt.rangeGrownBy(ir, 1)
        end
    end


function pair.static.rangeOfInnerStringSurroundingRange(lines, rng) 
        if belt.isInvalidLineIndex(lines, rng[2]) then return end
        
        local rgs = belt.rangesOfStringsInText(lines[rng[2]]:str(), rng[2])
        rgs = belt.rangesShrunkenBy(rgs, 1)
        for _, r in ipairs(rgs) do 
            if belt.rangeContainsRange(r, rng) then return r end
        end
    end


function pair.static.rangesOfStringsInText(text, li) 
        li = li or 1
        
        local rngs = array()
        local ss = -1
        local cc = nil
        for i in iter(1, #text) do 
            local c = string.sub(text, i, i)
            if (not cc and ((c == "'") or (c == '"'))) then 
                cc = c
                ss = i
            elseif (c == cc) then 
                if ((string.sub(text, (i - 1), (i - 1)) ~= '\\') or ((i > 2) and (string.sub(text, (i - 2), (i - 2)) == '\\'))) then 
                    rngs:push(array(ss, li, (i + 1), li))
                    cc = nil
                    ss = -1
                end
            end
        end
        
        return rngs
    end

--  0000000   000   000  000000000   0000000  000  0000000    00000000      0000000  000000000  00000000   
-- 000   000  000   000     000     000       000  000   000  000          000          000     000   000  
-- 000   000  000   000     000     0000000   000  000   000  0000000      0000000      000     0000000    
-- 000   000  000   000     000          000  000  000   000  000               000     000     000   000  
--  0000000    0000000      000     0000000   000  0000000    00000000     0000000      000     000   000  


function pair.static.positionsAndRangesOutsideStrings(lines, rngs, posl) 
        local found = array()
        
        for _, rng in ipairs(rngs) do 
            if not belt.isRangeInString(lines, rng) then 
                found:push(rng)
            end
        end
        
        for _, pos in ipairs(posl) do 
            if not belt.isRangeInString(lines, belt.rangeForPos(pos)) then 
                found:push(pos)
            end
        end
        
        return found
    end

-- 00000000    0000000   000  00000000    0000000         0000000  000   000  00000000   00000000   0000000     0000000   
-- 000   000  000   000  000  000   000  000             000       000   000  000   000  000   000  000   000  000        
-- 00000000   000000000  000  0000000    0000000         0000000   000   000  0000000    0000000    000   000  000  0000  
-- 000        000   000  000  000   000       000             000  000   000  000   000  000   000  000   000  000   000  
-- 000        000   000  000  000   000  0000000         0000000    0000000   000   000  000   000  0000000     0000000   


function pair.static.rangesOfNestedPairsAtPositions(lines, posl) 
        local rngs = array()
        for pos in posl:each() do 
            for pair in pepe.pairlAtCol(kseg.str(lines[pos[2]]), pos[1]) do 
                rngs:push(array(pair.rng[1], pos[2], pair.rng[2], pos[2]))
            end
        end
        
        return rngs
    end


function pair.static.spansOfNestedPairsAtPositions(lines, posl) 
        local spans = array()
        local brackets = array()
        local strings = array()
        
        for pos in posl:each() do 
            for pair in pepe.pairlAtCol(kseg.str(lines[pos[2]]), pos[1]) do 
                local open = array(pair.rng[1], pos[2], (pair.rng[1] + #pair.start))
                local close = array(pair.rng[2], pos[2], (pair.rng[2] + #pair.ende))
                spans:push(open)
                spans:push(close)
                
                if ((pair.start == '"') or (pair.start == "'")) then 
                    strings:push(open)
                    strings:push(close)
                else 
                    brackets:push(open)
                    brackets:push(close)
                end
            end
        end
        
        return array(spans, brackets, strings)
    end


function pair.static.rangesOfPairsSurroundingPositions(lines, pairl, posl) 
        local rngs = array()
        for pos in posl:each() do 
            for pair in pairl:each() do 
                if (kstr.endsWith(belt.chunkBeforePos(lines, pos), pair[1]) and kstr.startsWith(belt.chunkAfterPos(lines, pos), pair[2])) then 
                    rngs:push(array((pos[1] - #pair[1]), pos[2], (pos[1] + #pair[2]), pos[2]))
                end
            end
        end
        
        return rngs
    end


function pair.static.stringDelimiterSpansForPositions(lines, posl) 
        local spans = array()
        for pos in posl:each() do 
            local srng = belt.rangeOfStringSurroundingRange(lines, array(pos[1], pos[2], pos[1], pos[2]))
            if srng then 
                spans:push(array(srng[1], srng[2], (srng[1] + 1)))
                spans:push(array((srng[3] - 1), srng[4], srng[3]))
            elseif ((lines[pos[2]][pos[1]] == '"') or (lines[pos[2]][pos[1]] == "'")) then 
                srng = belt.rangeOfStringSurroundingRange(lines, array((pos[1] + 1), pos[2], (pos[1] + 1), pos[2]))
                if srng then 
                    spans:push(array(srng[1], srng[2], (srng[1] + 1)))
                    spans:push(array((srng[3] - 1), srng[4], srng[3]))
                end
            end
        end
        
        return spans
    end

--  ███████   ████████   ████████  ███   ███         ███████  ███       ███████    ███████  ████████
-- ███   ███  ███   ███  ███       ████  ███        ███       ███      ███   ███  ███       ███     
-- ███   ███  ████████   ███████   ███ █ ███        ███       ███      ███   ███  ███████   ███████ 
-- ███   ███  ███        ███       ███  ████        ███       ███      ███   ███       ███  ███     
--  ███████   ███        ████████  ███   ███         ███████  ███████   ███████   ███████   ████████


function pair.static.openCloseSpansForPositions(lines, posl) 
        local spans = array()
        for pos in posl:each() do 
            local sps = pair.openCloseSpansForPosition(lines, pos)
            if sps then 
                spans = spans + sps
            end
        end
        
        return spans
    end


function pair.static.openCloseSpansForPosition(lines, pos) 
        -- log "openCloseSpansForPosition #{lines}[ ]\npos #{pos}"
        
        local open = {
            ['['] = ']', 
            ['{'] = '}', 
            ['('] = ')'
            }
        
        local revs = {
            [']'] = '[', 
            ['}'] = '{', 
            [')'] = '('
            }
        
        local opns = dict.keys(open)
        local clos = dict.values(open)
        
        local maxLookups = 1000 -- careful, increasing this drops performance significantly!
        local lastOpen = nil
        local firstClose = nil
        local closeEncounters = ''
        local openEncounters = ''
        
        local bp = array(pos[1], pos[2])
        local stack = array()
        
        if not opns:has(lines[bp[2]][bp[1]]) then 
            closeEncounters = ''
            openEncounters = ''
            
            stack = array()
            
            local cnt = 0
            while true do 
                local cont = false
                bp[1] = bp[1] - 1
                if (bp[1] >= 1) then 
                    local prev = lines[bp[2]][bp[1]]
                    if opns:has(prev) then 
                        if (#stack > 0) then 
                            if (open[prev] == stack[#stack]) then 
                                openEncounters = openEncounters .. prev
                                stack:pop()
                                cont = true
                            else 
                                return
                            end
                        end
                        
                        if not cont then 
                            lastOpen = prev
                            break
                        end
                    elseif clos:has(prev) then 
                        stack:push(prev)
                        closeEncounters = closeEncounters .. prev
                    end
                else 
                    bp[2] = bp[2] - 1
                    if (bp[2] < 1) then break end
                    bp[1] = #lines[bp[2]]
                end
                
                if not cont then 
                    if lastOpen then break end
                    if (bp[2] < 1) then break end
                    cnt = cnt + 1
                    if (cnt > maxLookups) then break end
                end
            end
        else 
            lastOpen = lines[bp[2]][bp[1]]
        end
        
        stack = array()
        local ap = array(max((bp[1] + 1), pos[1]), pos[2])
        local cnt = 0
        
        while (ap[2] < lines:len()) do 
            local next = lines[ap[2]][ap[1]]
            if clos:has(next) then 
                if (#stack > 0) then 
                    if (open[stack[#stack]] == next) then 
                        stack:pop()
                    else 
                        return -- stack mismatch
                    end
                else 
                    firstClose = next
                    break
                end
            elseif opns:has(next) then 
                stack:push(next)
            end
            
            ap[1] = ap[1] + 1
            if (ap[1] > lines[ap[2]]:len()) then 
                ap[1] = 1
                ap[2] = ap[2] + 1
            end
            
            cnt = cnt + 1
            if (cnt > maxLookups) then 
                break
            end
        end
        
        -- log "lastOpen #{lastOpen} firstClose #{firstClose}" 
        if (not lastOpen or not firstClose) then 
            if (pos[2] > lines:len()) then print("pos[2] too large") ; return end
            if empty(lines[pos[2]]) then return end
            -- if pos[1]-1 > lines[pos[2]].len ➜ log "pos[1]-1 #{pos} too large #{lines[pos[2]].len}" ; ⮐  
            -- if lines[pos[2]].len > opns.len ➜ log "lines[pos[2]] #{lines[pos[2]].len} too large for revs #{revs.len}" ; ⮐  
            -- ⮐  if empty revs[lines[pos[2]]]
            -- if pos[1]-1 >= revs[lines[pos[2]]].len ➜ log "pos[1]-1 too large for revs" ; ⮐  
            if (clos:has(lines[pos[2]][(pos[1] - 1)]) and (kstr.find(openEncounters, revs[lines[pos[2]][(pos[1] - 1)]]) >= 1)) then 
                return pair.openCloseSpansForPosition(lines, array((pos[1] - 1), pos[2]))
            end
            
            return
        end
        
        -- log "lastOpen #{lastOpen} #{open[lastOpen]} firstClose #{firstClose}" 
        if (open[lastOpen] == firstClose) then 
            return array(array(bp[1], bp[2], (bp[1] + 1)), array(ap[1], ap[2], (ap[1] + 1)))
        end
        
        return
    end

return pair